### PR TITLE
feat/refactor(server): move CLI server start into API listen method

### PR DESCRIPTION
**Pull request by Loonride**
- [ ] This is a **bugfix**
- [x] This is a **feature**
- [x] This is a **code refactor**
- [ ] This is a **test update**
- [ ] This is a **docs update**
- [ ] This is a **metadata update**

### For Bugs and Features; did you add new tests?

Not yet, would like feedback on the idea first.

### Motivation / Use-Case

I have added a `start` method in the API, which the CLI calls, rather than calling the API's `listen` method.

- Firstly, the CLI will do significantly less configuration/error handling, as the API will do it instead
- API users can find more use in `start` method, because it does not make sense that they have to provide `host` and `port` via `listen(port, hostname, fn)`, when they probably already pass those into `new Server(options)` as dev server options.
- API users can make use of `socket` option and the automatic free port assignment via `start`, or opt out of those options via `listen`.

### Breaking Changes

Likely not, need to add tests to be sure though

### CODE REVIEW

File: bin/webpack-dev-server.js [hunk 1]
```diff
@@ -115,58 +112,7 @@ function startDevServer(config, options) {
...
-              }
-            });
-          }
-        });
-
-        clientSocket.connect({ path: options.socket }, () => {
-          throw new Error('This socket is already used');
-        });
-      }
-    });
```
Comment (evilebottnawi) [inline bin/webpack-dev-server.js]: I think we can solve this without new method, it is very inconsistently with node api for servers, we can move this code without new method, just listen `error` event and handle this inside server class

Comment (Loonride) [reply bin/webpack-dev-server.js]: Do you mean handle it inside the `listen(port, hostname, fn)` method, rather than making a new method?

**Comment by Loonride** [comment]
Tests that need to be added for this:

`socket` option:
- confirm that it works on CLI and API
- simulate `chmod` failure to ensure that handler works
- simulate EADDRINUSE for server start, then consequent ECONNREFUSED to see if the server can be created after initial failure

`port` option:
- should still work as string and number on API/CLI, if the string parses to a number
- on API, port should now go to a free port if an undefined one is provided

**Comment by Loonride** [comment]
> Let's separate this PR in two:
> 
> 1. Move socket check
> 2. Move findPort
> 
> Anyway looks good, better handle this in two PRs because:
> 
> 1. Easy review
> 2. Less chance of making a mistake
> 3. Easy create changelog
> 
> Anyway good job

Closing in favor of https://github.com/webpack/webpack-dev-server/pull/2061

**Pull request closed without merging**